Attempt to fix OSX libgcc linking issue.

pkgw opened the pull request:
The dynamic library that we produce on OSX has a busted dependency on a library named like `/Users/ray/mc-x86_64-3.5/.../libgcc_s.1.dylib`, which makes the resulting dylib unusable.

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipe```) and found it was in an excellent condition.

ocefpaf commented on recipe/meta.yaml:
```diff
@@ -20,19 +20,21 @@ source:
 requirements:
   build:
     - gcc  # [osx]
+    - libgcc  # [osx]
```
`gcc` installs `libgcc`, so this is not really needed (but it does not hurt to leave it there.)

pkgw replied:
Are you sure? `libgcc` wasn't pulled in during the first OSX build of this package, and the resulting library has a busted dependency on libgcc_s: [see this line of output from the OSX build](https://travis-ci.org/conda-forge/pgplot-feedstock/builds/196165092#L1046), and then look at what packages it installed at the beginning of the build.

ocefpaf replied:
Would the addition of `libgcc` at run time do it?

pkgw replied:
I tried installing it on my test machine and that didn't work.

ocefpaf replied:
Go for it then. But maybe @msarahan can chime in. Maybe the ``libgcc`` has something different that is not in the `gcc` packages that I am not aware of.

msarahan replied:
libgcc is potentially newer (could come from the GCC 5.2.0 docker image I made last year).  Installing them both at once is probably not a great idea - it's then ambiguous which files are present.  Whatever package got installed last...

pkgw replied:
Well, if you look at the build logs, the attempted fix didn't work — the resulting dylib still has a zany path for `libgcc_s` in the end. I'll need to do some more experimentation.

That being said, I think the `gcc` package is broken since it apparently contains a hardcoded path inside `/Users/ray` that shouldn't make it into a distributed package.

msarahan replied:
CC @mingwandroid

pkgw commented:
cc @mingwandroid @asmeurer 

Building on discussion from a previous iteration of this pull request that's now hidden, I've discovered that the bogus paths to `/Users/ray/mc-x64-3.5/conda-bld/.../libgcc_s.1.dylib` are embedded in the files `lib/libgcc_ext.10.{4,5}.dylib` in the `gcc` package. I'm currently getting version 4.8.5 build 7 of `gcc` from `defaults`.

*Update*: ... and I suspect this happens because `libgcc_ext` is a Mach-O "stub" file, which `install_name_tool` doesn't operate on; cf. [conda-build#140](https://github.com/conda/conda-build/issues/140) and  [conda-build#186](https://github.com/conda/conda-build/issues/186).

pkgw committed:
Fix the dylib on OSX.

As described in a new comment and the discussion on #1, the current gcc
package includes a dylib with a bad path that gets embedded in our
libpgplot.dylib. In this commit, we work around the problem. I also tidy up
the dependencies on libgfortran based on my improving understanding of how
these things all work. (The `gcc`, `libgcc`, and `libgfortran` packages all
provide overlapping sets of files, on OSX at least.)

We also get rid of some pkg-config invocations that don't work and aren't
needed.

mingwandroid commented:
`lib/libgcc_ext.10.{4,5}.dylib` do not contain that string. They contain:

```
Load command 3
          cmd LC_ID_DYLIB
      cmdsize 304
         name /Users/ray/mc-x64-3.5/conda-bld/gcc-4.8_1477649012852/_b_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_plac/lib/libgcc_s.1.dylib (offset 24)
```

.. which should be fine, but stub dylibs are fairly undocumented things. I think to get on the same page as you I would need to attempt to build this.

pkgw commented:
Just for clarity, the `...` above is me ellipsizing the long placeholder string, not the literal text that I was seeing.

For me, using the `gcc` package's `gfortran` to create any dylib results in objects that embed the nonexistent `/Users/ray/...` path. It looks like regular executables can link to the resulting dylibs and be run successfully. The problem that I saw was that Python 2.7 would fail to load extension modules linking to such dylibs ... perhaps the dynamic linker is pickier in those situations.

mingwandroid commented:
It's a stub library and that string appears in the `LC_DYLIB_ID` command so there should not be a problem here.

Edit: `LC_DYLIB_ID` is just how dylibs identify themselves to dyld, they are not `LC_LOAD_DYLIB` commands. Those are the ones that contain paths that need to be found.

mingwandroid commented:
@pkgw, can you try this instead?

```
if [ -n "$OSX_ARCH" ] ; then
    subtype=actually_osx

    # install_name_tool refuses to operate on stub dylibs, so binary patch the
    # LC_DYLIB_ID name instead.
    for name in gcc_ext.10.4 gcc_ext.10.5 ; do
        OLD="$(echo -n /Users/ray/mc-x64-3.5/conda-bld/gcc-4.8_1477649012852/_b_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_plac/lib/libgcc_s.1.dylib | xxd -g 0 -u -ps -c 2560)00"
        NEW="$(echo -n $PREFIX/lib/libgcc_s.1.dylib | xxd -g 0 -u -ps -c 2560)00"
        hexdump -ve '1/1 "%.2X"' $PREFIX/lib/lib${name}.dylib | \
        sed "s|${OLD}|${NEW}|g" | \
        xxd -r -p > $PREFIX/lib/lib${name}.dylib.patched
        rm ${PREFIX}/lib/lib${name}.dylib
        mv ${PREFIX}/lib/lib${name}.dylib.patched ${PREFIX}/lib/lib${name}.dylib
        done
else
    subtype=gfortran_gcc
fi
```

(This is hopefully safer since stub dylibs are actually a lot more than just a reference to another dylib, they provide implementations of the stubbed-through-to-functions incase the real dylib doesn't contain them).

mingwandroid commented:
Here's the end of my build log:

```
pgplot
------

conda-forge::libgfortran-3.0.0-0:
    libgcc_s.1.dylib (lib/libgcc_s.1.dylib)
    libgfortran.3.dylib (lib/libgfortran.3.dylib)
    libquadmath.0.dylib (lib/libquadmath.0.dylib)

conda-forge::libpng-1.6.28-0:
    libpng16.16.dylib (lib/libpng16.16.dylib)

conda-forge::xorg-libx11-1.6.4-6:
    libX11.6.dylib (lib/libX11.6.dylib)

local::pgplot-5.2.2-1:
    libpgplot.dylib (lib/libpgplot.dylib)

system:
    libSystem.B.dylib (/usr/lib/libSystem.B.dylib)

not found:


+ conda inspect objects -p /Users/rdonnelly/cf/conda-bld/pgplot_1487553891698/_t_env pgplot
pgplot
------

bin/pgxwin_server
  filetype: EXECUTE
  rpath: @loader_path/../lib/


lib/libpgplot.dylib
  filetype: DYLIB
  rpath: /Users/rdonnelly/cf/conda-bld/pgplot_1487553891698/_t_env/lib:@loader_path/



TEST END: pgplot-5.2.2-1
```

Next, I am going to try a build without hacking the `LC_DYLIB_ID` name as I suspect it's unnecessary.

mingwandroid commented:
Without these hacks I do get:

```
Load command 14
          cmd LC_LOAD_DYLIB
      cmdsize 304
         name /Users/ray/mc-x64-3.5/conda-bld/gcc-4.8_1477649012852/_b_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_plac/lib/libgcc_s.1.dylib (offset 24)
   time stamp 2 Thu Jan  1 01:00:02 1970
      current version 1.0.0
compatibility version 1.0.0
```

in `lib/libpgplot.dylib`, but it appears to be harmless, I'm going to speculate that Apple's `dyld` replaces references to `libgcc_s.1.dylib` with references to `libSystem.B.dylib`.

Can you point me at a Python module that cannot be loaded though? I need to fix that somehow.

I'll play around with this some more tomorrow.

pkgw commented:
I'm having trouble reproducing an example module that fails to load, since my use case involves a large package with a nasty web of dependencies (namely, [this](https://anaconda.org/pkgw-forge/casa-python)). I'll see what I can do, though.

mingwandroid commented:
Can you try with https://github.com/conda-forge/pgplot-feedstock/pull/1#issuecomment-280965748

If it works for you then I'll probably move it into a post link script in GCC.

mingwandroid commented:
I fixed this 'properly' now in default's macOS gcc-4.8.5-8 package (as properly as I could anyway), your hack should not be required.
